plugins/ipfs/util.go

```diff
@@ -0,0 +1,310 @@
+	"github.com/multiformats/go-multiaddr"
+	"github.com/pkg/errors"
+
+	"github.com/ipfs/iptb/testbed/interfaces"
+)
+
+type ListenType int
+
+const (
+	attrID   = "id"
+	attrPath = "path"
+
+	metricBwIn  = "bw_in"
+	metricBwOut = "bw_out"
+
+	LT_TCP ListenType = iota
+	LT_WS
+	LT_UTP
+)
+
+func InitIpfs(l testbedi.Core) error {
+	return nil
+}
+
+func SwarmAddr(lt ListenType, addr string, port int) string {
+	str := "/ip4/%s/tcp/%d"
+
+	switch lt {
+	case LT_WS:
+		str = "/ip4/%s/tcp/%d/ws"
+		break
+	case LT_UTP:
+		str = "/ip4/%s/udp/%d/utp"
+		break
+	}
+
+	return fmt.Sprintf(str, addr, port)
+}
+
+func GetAttr(l testbedi.Core, attr string) (string, error) {
+	switch attr {
+	case attrID:
+		pcid, err := l.PeerID()
+		if err != nil {
+			return "", err
+		}
+		return pcid, nil
+	case attrPath:
+		return l.Dir(), nil
+	default:
+		return "", errors.New("unrecognized attribute: " + attr)
+	}
+}
+
+func GetMetric(l testbedi.Core, metric string) (string, error) {
+	switch metric {
+	case metricBwIn:
+		bw, err := GetBW(l)
+		if err != nil {
+			return "", err
+		}
+		return fmt.Sprint(bw.TotalIn), nil
+	case metricBwOut:
+		bw, err := GetBW(l)
+		if err != nil {
+			return "", err
+		}
+		return fmt.Sprint(bw.TotalOut), nil
+	default:
+		return "", errors.New("unrecognized metric: " + metric)
+	}
+}
+
+func GetPeerID(l testbedi.Config) (*cid.Cid, error) {
+	icfg, err := l.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	lcfg, ok := icfg.(*config.Config)
+	if !ok {
+		return nil, fmt.Errorf("Error: GetConfig() is not an ipfs config")
+	}
+
+	pcid, err := cid.Decode(lcfg.Identity.PeerID)
+	if err != nil {
+		return nil, err
+	}
+
+	return pcid, nil
+}
+
+func GetMetricList() []string {
+	return []string{metricBwIn, metricBwOut}
+}
+
+func GetMetricDesc(metric string) (string, error) {
+	switch metric {
+	case metricBwIn:
+		return "node input bandwidth", nil
+	case metricBwOut:
+		return "node output bandwidth", nil
+	default:
+		return "", errors.New("unrecognized metric")
+	}
+}
+
+func GetAttrList() []string {
+	return []string{attrID, attrPath}
+}
+
+func GetAttrDesc(attr string) (string, error) {
+	switch attr {
+	case attrID:
+		return "node ID", nil
+	case attrPath:
+		return "node IPFS_PATH", nil
+	default:
+		return "", errors.New("unrecognized attribute")
+	}
+}
+
+func ReadLogs(l testbedi.Libp2p) (io.ReadCloser, error) {
+	addrStr, err := l.APIAddr()
+	if err != nil {
+		return nil, err
+	}
+
+	addr, err := multiaddr.NewMultiaddr(addrStr)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO(tperson) ipv6
+	ip, err := addr.ValueForProtocol(multiaddr.P_IP4)
+	if err != nil {
+		return nil, err
+	}
+	pt, err := addr.ValueForProtocol(multiaddr.P_TCP)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := http.Get(fmt.Sprintf("http://%s:%s/api/v0/log/tail", ip, pt))
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.Body, nil
+}
+
+type BW struct {
+	TotalIn  int
+	TotalOut int
+}
+
+func GetBW(l testbedi.Libp2p) (*BW, error) {
```

Comment:
Might as well take a context for all of these functions up-front (unless I'm missing something about how they're supposed to be used).